### ISSUE TITLE
Proposal: Using sub-types of an interface or class for typing

### ISSUE THREAD

**Issue by renehamburger** [Microsoft/TypeScript#12481]
There are many scenarios where complex type hierarchies (interfaces/classes) are needed, for example when typing the response of an API or the options object of a JavaScript library. But it can be quite cumbersome to define each segment/sub-type in such a hierarchy separately in order to use it as a type elsewhere.

It is already possible to use the type of any nested property of an instance for typing:

```ts
interface Response { 
    user: {
        id: string;
        name: string;
    };
    data: any;
}

const r: Response = null;

const user: typeof r.user = {
    id: '1',
    name: 'SomeUser'
};
```

But it would be convenient to do that without the need of an instance:

```ts
interface Response { 
    user: {
        id: string;
        name: string;
    };
    data: any;
}

const user: Response.user = {
    id: '1',
    name: 'SomeUser'
};
```

In this case it would be nice to have and save a lot of code if the interface has a very deep hierarchy. But it's even more important for type hierarchies declared in external definition files. See for example the Options interface defined for [handsontable](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/types-2.0/handsontable/index.d.ts#L129):

```ts
interface Options {
  ...
  // hooks
  afterAutofillApplyValues?: (startArea: any[], entireArea: any[]) => void;
  afterCellMetaReset?: () => void;
  afterChange?: (changes: any[], source: string) => void;
  afterChangesObserved?: () => void;
  afterColumnMove?: (startColumn: number, endColumn: number) => void;
  afterColumnResize?: (currentColumn: number, newSize: number, isDoubleClick: boolean) => void;
  afterColumnSort?: (column: number, order: boolean) => void;
   ...
}
```

None of these callbacks are defined separately, so it would be great to be able to use them directly with `const cb: Options.afterAutofillApplyValues;`.

(This may well have been proposed before, but I couldn't find anything in the current issue list.)

**Comment by yortus** [comment]
Looks like a duplicate of #4555, which was closed in favour of #4640 which was closed in favour of #6606.

Although the proposal in #6606 addresses this issue, the [guidance from the team](https://github.com/Microsoft/TypeScript/issues/6606#issuecomment-178266654) about what PR would be accepted does *not* cover it. Perhaps someone from the team can clarify whether anything has changed in terms of supporting this.

**Comment by yortus** [comment]
Note [this comment](https://github.com/Microsoft/TypeScript/issues/6606#issuecomment-260154482) in #6606 shows something similar that works in the current nightly.
